ignite/metrics/epoch_metric.py

```diff
@@ -136,31 +135,32 @@ def update(self, output: Tuple[torch.Tensor, torch.Tensor]) -> None:
             except Exception as e:
                 warnings.warn(f"Probably, there can be a problem with `compute_fn`:\n {e}.", EpochMetricWarning)
 
-    def compute(self) -> Any:
+    def compute(self) -> float:
         if len(self._predictions) < 1 or len(self._targets) < 1:
             raise NotComputableError("EpochMetric must have at least one example before it can be computed.")
 
-        _prediction_tensor = torch.cat(self._predictions, dim=0)
-        _target_tensor = torch.cat(self._targets, dim=0)
+        if not self._is_reduced:
+            _prediction_tensor = torch.cat(self._predictions, dim=0)
+            _target_tensor = torch.cat(self._targets, dim=0)
 
-        ws = idist.get_world_size()
+            ws = idist.get_world_size()
+            if ws > 1:
+                # All gather across all processes
+                _prediction_tensor = cast(torch.Tensor, idist.all_gather(_prediction_tensor))
+                _target_tensor = cast(torch.Tensor, idist.all_gather(_target_tensor))
 
-        if ws > 1 and not self._is_reduced:
-            # All gather across all processes
-            _prediction_tensor = cast(torch.Tensor, idist.all_gather(_prediction_tensor))
-            _target_tensor = cast(torch.Tensor, idist.all_gather(_target_tensor))
-        self._is_reduced = True
+            self._result = 0.0
```

Comment:
I think we should define `self._result = None` in `reset` method and use `if self._result is None` to run result computations. Also `reinit__is_reduced` should set `self._result = None` if attribute exists.

Reply:
Why not to use `self._is_reduced` in the `if` statement in `EpochMetric`? Anyway we have to set it to `True` at the end of the statement.

Reply:
IMO, `self._is_reduced` is not applicable here and everywhere once https://github.com/pytorch/ignite/pull/2803 is merged. The role of `self._is_reduced` was to avoid unnecessary (even faulty) dist reductions as we were overwriting internal counters. In #2803 we find out that this way of computing metrics has a bug. I think we should avoid using `self._is_reduced` at all